jgrapht-core/src/test/java/org/jgrapht/alg/vertexcover/VertexCoverTest.java

```diff
@@ -0,0 +1,30 @@
+/*
+ * (C) Copyright 2018-2018, by Linda Buisman and Contributors.
+ *
+ * JGraphT : a free Java graph-theory library
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either
+ *
+ * (a) the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation, or (at your option) any
+ * later version.
+ *
+ * or (per the licensee's choosing)
+ *
+ * (b) the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation.
+ */
+package org.jgrapht.alg.vertexcover;
+
+import org.jgrapht.alg.interfaces.MinimumVertexCoverAlgorithm;
+
+/**
+ * Tests the vertex cover algorithms.
+ *
+ * @author Linda Buisman
+ */
+public interface VertexCoverTest extends VertexCoverTestBase {
```

Comment:
Do we need this additional interface? Can't we simply have an abstract class VertexCoverTestBase with abstract method 
`<V, E> MinimumVertexCoverAlgorithm<V, E> createSolver();` and have the other algorithms extend `VertexCoverTestBase` or am I missing something obvious here?

Reply:
We need a `createSolver` which returns a `MinimumVertexCoverAlgorithm` and another one which returns a `MinimumWeightedVertexCoverAlgorithm` (for the weighted case).

I want to be able to test an implementation that only works in the unweighted case (without changing anything in other classes). And I also want to have only one test class per implementation (i.e. when I run `BarYehudaEvenTwoApproxVCImplTest` I want to run all tests, both weighted and unweighted). Since Java doesn't support multiple inheritance, I had to use two interfaces.

The structure of this hierarchy:
 - VertexCoverTestBase: interface which contains useful methods (e.g. `isCover`)
 - VertexCoverTest: interface for unweighted tests; contains `<V, E> MinimumVertexCoverAlgorithm<V, E> createSolver();`
 - WeightedVertexCoverTest: interface for weighted tests; contains `<V, E> MinimumWeightedVertexCoverAlgorithm<V, E> createWeightedSolver();`
 - `VertexCoverExactTest`, `VertexCoverTwoApproxTest`: abtract classes which implement `VertexCoverTest` and contain the actual tests
 - `WeightedertexCoverExactTest`, `WeightedVertexCoverTwoApproxTest`: abtract classes which extend the corresponding unweighted versions and implement `WeightedVertexCoverTest` and contain the actual tests
 - `EdgeBasedTwoApproxVCImplTest` is the only concrete class which extends `VertexCoverTwoApproxTest` (because `EdgeBasedTwoApproxVCImpl<V, E>
    implements MinimumVertexCoverAlgorithm<V, E>` and not `MinimumWeightedVertexCoverAlgorithm<V, E>` like all other VC implementations available)

Reply:
I would try to simplify this and avoid using base interfaces with fields and default methods for code reuse.

I would do the following: 
 - Methods `isCover()` and `createRandomPseudograph()` can be added in a helper static class. 
 - I would separate the two hierachies, one for weighted and one for unweighted.
 - Remove the base interface.